### PR TITLE
Feature: Canvas element drag and drop with snap to grid

### PR DESCRIPTION
**Pull request by rgerstenberger**
This adds snap to grid functionality for canvas elements and many other improvements to drag and drop.

snap-to-grid

/cc @mdaxtman @kenwheeler

### CODE REVIEW

**Comment by kenwheeler** [comment]
looks good to me (ci failure aside)

File: app/utils/index.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
+import { sortBy, sortedUniq } from "lodash";
+
+import { ElementTypes, SNAP_DISTANCE } from "../constants";
+
+export const getElementDimensions = ({ type, props }) => {
+  if (type === ElementTypes.TEXT) {
+    return {
```
Comment (mdaxtman) [inline app/utils/index.js]: Do we want white space here?

Comment (rgerstenberger) [inline app/utils/index.js]: Nah, since it's the only statement in the block.

File: app/utils/index.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+// NOTE: Because this depends on a hash with integer keys, all potentialLines must be integers
+export const getNearestGridLine = (gridLinesHash, ...potentialLines) => {
+  let line = gridLinesHash[Math.floor(potentialLines[0])];
+  let distance = Math.abs(line - Math.floor(potentialLines[0]));
+  let index = 0;
+
+  for (let i = 1; i < potentialLines.length; i++) {
+    const currentLine = gridLinesHash[Math.floor(potentialLines[i])];
+    const currentDistance = Math.abs(currentLine - Math.floor(potentialLines[i]));
+    if (currentDistance < distance) {
```
Comment (mdaxtman) [inline app/utils/index.js]: or here?

**Comment by mdaxtman** [comment]
LGTM too.

File: app/utils/index.js (hunk 2 above)
Comment (rgerstenberger) [inline app/utils/index.js]: Here yes, good catch!

**Pull request merged**